src/processes/poisson.jl

```diff
@@ -19,61 +24,94 @@ Base.union(p₁::PoissonProcess{<:Real}, p₂::PoissonProcess{<:Function}) = Poi
+default_sampling_algorithm(::PoissonProcess) = DiscretizedSampling()
+default_sampling_algorithm(::PoissonProcess{<:Function}) = SimpleThinnedSampling()
 
 #------------------
 # HOMOGENEOUS CASE
 #------------------
 
-function rand_single(rng::Random.AbstractRNG, p::PoissonProcess{<:Real}, g::Geometry, ::DiscretizedSampling)
+function rand_single(rng::Random.AbstractRNG, p::PoissonProcess{<:Real}, g, ::DiscretizedSampling)
   # simulate number of points
   λ = p.λ
   V = measure(g)
   n = rand(rng, Poisson(λ * V))
 
-  # simulate homogeneous process
-  pts = sample(g, HomogeneousSampling(n))
+  if n == 0
+    # PointSet{embeddim(g), coordtype(g)}([])
+    nothing
+  else
+    # simulate homogeneous process
+    pts = sample(g, HomogeneousSampling(n))
 
-  # return point pattern
-  PointSet(collect(pts))
+    # return point pattern
+    PointSet(collect(pts))
+  end
 end
 
-function rand_single(rng::Random.AbstractRNG, p::PoissonProcess{<:Real}, b::Box, ::ProductSampling)
-  # region configuration
-  lo, up = coordinates.(extrema(b))
+#--------------------
+# INHOMOGENEOUS CASE
+#--------------------
 
+function rand_single(rng::Random.AbstractRNG, p::PoissonProcess{<:Vector}, m::Mesh, algo::DiscretizedSampling)
   # simulate number of points
   λ = p.λ
-  V = measure(b)
-  n = rand(rng, Poisson(λ * V))
+  V = measure.(m)
+  n = rand(rng, Poisson(sum(λ .* V)))
+
+  if n == 0
+    nothing
+  else
+    # sample elements with weights proportial to expected number of points
+    w = WeightedSampling(n, λ .* V, replace = true)
+
+    # within each element sample a single point
+    h = HomogeneousSampling(1)
+    pts = (first(sample(rng, e, h)) for e in sample(rng, m, w))
+
+    # return point pattern
+    PointSet(collect(pts))
+  end
+end
 
-  # product of uniform distributions
-  U = product_distribution([Uniform(lo[i], up[i]) for i in 1:embeddim(b)])
+function rand_single(rng::Random.AbstractRNG, p::PoissonProcess{<:Function}, g, algo::ThinnedSampling)
+  # simulate a homogeneous process
+  pp = rand(rng, PoissonProcess(algo.λmax), g)
 
-  # return point pattern
-  PointSet(rand(rng, U, n))
+  # thin point pattern
+  thin(pp, RandomThinning(x -> p.λ(x) / algo.λmax))
 end
 
-#--------------------
-# INHOMOGENEOUS CASE
-#--------------------
-
-function rand_single(rng::Random.AbstractRNG, p::PoissonProcess{<:Function}, b::Box, algo::DiscretizedSampling)
-  # discretize region
-  # TODO
+function rand_single(rng::Random.AbstractRNG, p::PoissonProcess{<:Function}, g, algo::SimpleThinnedSampling)
+  # compute intensity on vertices
+  box = boundingbox(g)
+  if isnothing(algo.dims)
+    m = CartesianGrid(box.min, box.max)
+  else
+    m = CartesianGrid(box.min, box.max, dims = algo.dims)
+  end
+  v = vertices(m)
+
+  # obtain λmax
+  λvec = p.λ.(v)
+  λmax = maximum(λvec) + 0.05 * (maximum(λvec) - minimum(λvec))
+
+  # simulate inhomogeneous process
+  rand_single(rng, p, g, ThinnedSampling(λmax))
+end
```

Comment:
I understand that this is a procedure to find a good lambda_max for a grid. I would simply introduce a new function to get this default_lambda_max and would use it elsewhere. Right now we are artificially introducing a new method of rand_single just to compute the lambda_max.

Also, maybe we can remove the redundancy between ThinnedSampling and SimpleThinnedSampling with a type parameter? It is not clear to me yet why we need two separate structs.